lib/pbench/server/api/resources/query_apis/datasets/table_of_content.py

```diff
@@ -74,22 +91,61 @@ def assemble(self, json_data: JSON, context: CONTEXT) -> JSON:
         }
 
     def postprocess(self, es_json: JSON, context: CONTEXT) -> JSON:
+        """
+        Returns a Flask Response containing a JSON object (keyword/value
+        pairs) where each key is the fully qualified name of a keyword
+        (sub-)field (from run-toc documents) and corresponding value
+        is a non-empty value or list of values which appear in that field.
+
+        Example:
+
+            {
+                "_index": "riya-pbench.v6.run-toc.2021-05",
+                "_type": "_doc",
+                "_id": "3bba25b62fhdgfajgsfdty6797ed06a",
+                "_score": 0.0,
+                "_source": {
+                    "parent": "/default",
+                    "directory": "/default/sample",
+                    "mtime": "2021-05-21T24:00:00",
+                    "mode": "0o755",
+                    "name": "sample",
+                    "ancestor_path_elements": ["default"],
+                    "files": [
+                        {
+                            "name": "result.txt",
+                            "mtime": "2021-05-21T24:00:00",
+                            "size": 0,
+                            "mode": "0o644",
+                            "type": "reg",
+                        },
+                        {
+                            "name": "user-benchmark.cmd",
+                            "mtime": "2021-05-21T24:00:00",
+                            "size": 114,
+                            "mode": "0o755",
+                            "type": "reg",
+                        },
+                    ],
+                    "run_data_parent": "ece030b7b37ajsdgfa66s6a7a6be804",
+                    "authorization": {"owner": "1", "access": "private"},
+                    "@timestamp": "2021-05-21T24:00:00.141295",
+                },
+            }
+        """
         dir_list = []
         file_list = []
         for toc in es_json["hits"]["hits"]:
             directory = toc["_source"]["directory"]
             parent = toc["_source"]["parent"]
             direct = directory.rsplit("/", 1)[0]
             try:
-                if (
-                    parent != directory
-                ):  # This condition is especially when parent and directory = '/'
-                    t = toc["_source"]["name"]
-                    dir_list.append(t)
-                if not file_list:
-                    if parent == directory or parent == direct:
-                        file_list = toc["_source"]["files"]
-            except KeyError:
-                pass
+                if directory != "/":
+                    dir_list.append(toc["_source"]["name"])
+                if not file_list and (parent == directory or parent == direct):
+                    file_list = toc["_source"]["files"]
+            except KeyError as exc:
+                if exc.args[0] == ["files", "name"]:
+                    pass
```

Comment:
> Do lines 143-144 really need to be inside the `try` block? (That is, do we really have cases where `"name"` is missing?)

yes, we do have conditions where `name` value is not there. It also comes in need where there is nothing inside a directory

>Also, is the test for not file_list in 145 really necessary?

It is also a requirement because there are some cases, where there are multiple sub-directories and there are multiple subdirectories which fulfil thelater condition but we don't want the overriding

>it gets us the first one that we find, and how are we to know that it's better than the second!?

this is something to consider.

Reply:
> Finally, for line 146, would it make sense to use `toc["_source"].get("files", [])` to avoid the need for the `try` block? (If so, perhaps we could do something somewhat similar for `"name"` as well and get rid of the `try` block altogether.)

this seems to be working :) will update it